### ISSUE TITLE
Inefficient rendering of views

### ISSUE THREAD

**Issue by jamesplease** [jridgewell/minionette#10]
By my count, it seems like there could be up to `4n + 1` DOM operations whenever a view renders itself, where `n` is the number of regions in the view.
1. Detach the current view in the region, if it exists (for every `n`)
2. Reattach the region's element (for every `n`)
3. Render, then set, the view's html
4. Attach the old view from the region (if it exists) (for every `n`)
5. Detach the region's element (for every `n`)

2 and 5 on this list may not be all that bad, given the default tag of regions, but 1, 3, and 4 could potentially cause reflows.

A more efficient, though also more complex, implementation, might look something like:
1. Render the template, but don't attach it
   
   `var html = template(data);`
2. Convert it to a DOM tree
   
   `var tree = $.parseHTML(html)`
   
   Another option would be using [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser), depending on your target browser compatibility.
3. Clone the views in each of the regions, and inject them into the newly-created tree
4. Replace the contents of this view's element with this new tree

This would reduce the # of DOM operations from `4n + 1` to 2.

**Comment by jridgewell** [comment]
:+1:. I'll look into this for v2.

**Comment by jamesplease** [comment]
Notes:
- It's far easier to search through children of a jquery object than the matched elements+children, so using a wrapper element is handy
  
  ``` js
  var htmlString = '<div>' + renderedTemplateHtml + '</div>';
  var $domTree = $($.parseHTML(htmlString));
  ```
- You can remove the wrapper at the end of the operation via `.contents()`
  `this.$el.empty().append($domTree.contents());`
  These are the 2 necessary DOM operations I referred to in the original post
- The code for finding and replacing is pretty simple, really, if you give your regions an ability to clone their tree:
  `$domTree.find(region.selector).replaceWith(region.cloneTree());`
  You simply loop through your regions and run that code. Also, this line is the reason the wrapper element is convenient.

This is a first pass, and it seems to be working well for me, but I need to do some perf tests.

~~I'm going to see if cloning the elements is even necessary. I see no reason why we can't simply reuse the same DOM tree as before!~~

**Update:** cloning _is_ required. However, cloning causes the children to lose references to the elements actually being displayed. So there needs to be some sort of recursive reassigning of elements. Exploring further now!

**Update2:** Solved [over here](https://github.com/jmeas/puppets/tree/v0.0.1/src) (in the region and view files)

**Comment by jridgewell** [comment]
I'm not too partial to recursively updating `$el` references (especially because `#setElement` is going to cause [havoc with events](https://github.com/jashkenas/backbone/blob/master/backbone.js#L1123-L1126)).

I do think re-rendering can be more efficient, though. How about:
1. Create new DOM tree for top view.
   
   ``` javascript
   // is $.parseHTML necessary? Won't just `$(...)` suffice?
   var $el = $('<div>' + template() + '</div>');
   ```
2. Loop through the regions, replacing each placeholder selector in the new `$el` with the region's view's `$el`.
   
   ``` javascript
   _.each(this._regions, function(region) {
       $el.find(region.selector).replaceWith(region.view.$el);
   });
   ```
3. Replace the top view's `$el`.
   
   ``` javascript
   this.$el.empty().append($el.contents());
   ```

This'll still cause `n + 1` repaints. But, this'll keep any `region.view.$el` references alive and won't cause any event mayhem.

And, it'll eliminate the need for placeholder views. :smile:

**Comment by jridgewell** [comment]
Spitballing:

``` javascript
function render() {
    this.$el.contents().detach()
    var $el = $('<div>').append(template());
    _.each(this._regions, function(region) {
        $el.find(region.selector).replaceWith(region.view.$el);
    });
    this.$el.append($el.contents());
}
```

That's two repaints, correct?

---

On the topic of Regions:

I think the ~~most~~ _only_ important aspect is the ability to attach a new view in place of the current one. What if we add `View#replaceWith(newView, remove = true)`? Wouldn't that cover 99% of the need for a Region?

**Comment by jamesplease** [comment]
> I'm not too partial to recursively updating $el references (especially because #setElement is going to cause havoc with events).

`#_setElement` causes havoc, but when you use `setElement` it saves any events configured in the `events` hash. For now, this is the approach I've taken. With that said, you can pass `true, true` [to clone](http://api.jquery.com/clone/) to copy over **everything** – all data and events, which would let us do a straight set `this.$el = $cloned; this.el = $cloned[0]`. I had this originally, then removed it, and now I'm undecided due to perf concerns.

This whole idea of mine is a big experiment that basically relies on the fact that big jQuery operations will be more preferable than devastating DOM operations. But I haven't ever done jQuery ops like this on potentially-large DOM trees. I'm not sure at what point the system will break down!

> // is $.parseHTML necessary? Won't just `$(...)` suffice?

Heh yeah, probably. I should make a JSPerf to see how they compare perf-wise.

> And, it'll eliminate the need for placeholder views. 

I don't think that placeholder views have ever been necessary? I never understood that part of Minionette's regions. As I see it, regions just need to be references to elements that serve as the root of views, which can then be swapped in and out. If there's no view, then it's still just an element, and that's okay.

> I think the most only important aspect is the ability to attach a new view in place of the current one.

I'm keeping them around for four reasons:
1. The reason you mentioned: easily swappable views.
2. Solves the view wrapper problem by making the element a node that is already in another view's template. I much prefer this over LayoutManager / Ampersand's approach to making the wrapper optional. There's way less logic, for instance.
3. They provides a nested structure that can be serialized, and presented in something like [the Inspector](https://github.com/marionettejs/marionette.inspector). Simply using jQuery attach or replace to compose view trees is just throwing away valuable data about the structure of your view tree. If we got rid of regions and only made views swappable, then I would want a new way to codify the nesting. Maybe a wrapper around jQuery.append/replaceWith?
4. They will power my animation library, which will be a critical reason to use my system over, say, Marionette, which provides 0 abstraction for animation. In Marionette, you essentially have to go back to jQuery animations, which isn't the worst, but we can do better. That 'better' is stealing the source of `ngAnimate`, which is what also powers React's animations. This might be possible w.o. regions though.

# 

> That's two repaints, correct?

I tried something similar to this, I think. If I was testing the same thing, it removed the elements from the page when setting them in the newly created tree, which caused a whole bunch of DOM ops.

**Comment by jridgewell** [comment]
> With that said, you can pass true, true to clone to copy over everything – all data and events, which would let us do a straight set this.$el = $cloned; this.el = $cloned[0]. I had this originally, then removed it, and now I'm undecided due to perf concerns.

You would have to do this, since any custom events won't be copied over with `#setElement`.

> This whole idea of mine is a big experiment that basically relies on the fact that big jQuery operations will be more preferable than devastating DOM operations. But I haven't ever done jQuery ops like this on potentially-large DOM trees. I'm not sure at what point the system will break down!

I'm on board with this, I think'll be better myself. I'm going to write a bench test using this page as an example. Treat each comment as a separate region, and go through Minionette's way vs. yours. I'm pretty sure yours'll win handily.

> I don't think that placeholder views have ever been necessary? I never understood that part of Minionette's regions.

It's essentially legacy code, but one that I personally like. The original Region implementation didn't use selectors, it used the `view` template helper. It would create a new region if it didn't exist, and insert the placeholder view in that spot of the template so that code could later replace it. It also played very nicely with the selector implementation.

> Solves the view wrapper problem...

Do you mean the wrapping element around the entire subview, like Marionettes? I wasn't thinking of making that the case.

> They provides a nested structure that can be serialized ... Simply using jQuery attach or replace to compose view trees is just throwing away valuable data about the structure of your view tree.

Sorry, it didn't explain my meaning very well. I mean something more like:

``` javascript
var view = new View({
    regions: {
        'content': '.selector'
    }
});
// placeholder
view.content instanceof Backbone.View; // => true
view.content.replaceWith(newView);
view.content === newView; // => true

view = new View();
view.content; // => undefined
view.addRegion('.selector', newView);
view.content === newView; // => true
```

So the parent view would still have all the references to it's children. Really, it could be more like you're proposed Marionette methods, too, though that'd mean dropping support for the `view` helper.

``` javascript
view.showChildElement('.selector', newView);
view.selector === newView; // => true
view.showChildElement('.selector', otherView);
view.selector === otherView; // => true
```

I need to think on this one some more.

> They will power my animation library

I have zero animation expertise outside the normal jQuery stuff, so I'll have to see with what you come up with here. I'll look into `ngAnimate`, though.

---

> it removed the elements from the page when setting them in the newly created tree, which caused a whole bunch of DOM ops.

Doesn't seem to be any more than `.empty().append($.contents())`.

![#empty and #append](https://dl.dropboxusercontent.com/spa/keyy7k9vgnogrqa/s4xiitki.png)

And here's using my proposed way:

![#append and #append](https://dl.dropboxusercontent.com/spa/keyy7k9vgnogrqa/on_q7cfi.png)

**Comment by jridgewell** [comment]
Here are my results after testing:
1. Minionette's current way: Takes roughly [883ms to rerender](https://www.dropbox.com/s/u1npvyzybeivk87/minionette-timeline.json?dl=0).
2. Your proposed clone rerender: Takes roughly [610ms](https://www.dropbox.com/s/kzwnd3ojom6bb5r/clone-html-append-timeline.json?dl=0)
3. My `$el.contents().detach()` rerender: Takes roughly [429ms](https://www.dropbox.com/s/i5vdwlnhhcja2hn/html-detach-append-timeline.json?dl=0)

Even when I disabled the cloning of jQuery events, your clone way still took 582ms.

Here's my test setup: https://www.dropbox.com/s/s0fdmpljhzh3ims/dom-test.zip?dl=0. I had to inflate the number of regions to really see the results. Each one was within 10ms for me when I didn't.

**Comment by jridgewell** [comment]
[MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) from https://github.com/jmeas/puppets/issues/6.

**Comment by jamesplease** [comment]
@jridgewell, thanks for the tests! I'm still checking them out.

In the meantime, I made a test repo that I plan to use as the basis for my testing while writing Puppets. You can check it out [here](https://github.com/jmeas/dom-op-tester).

For now, there are 3 tests.

The first is a basic test showing that emptying an element then setting its text causes 2 DOM ops. It's just something to use as an example to write your own tests.

The second is a test case showing what I think is your proposed method, which causes `n+1` DOM ops.

The third is a test case showing my proposed method, which causes a single DOM op.

> I had to inflate the number of regions to really see the results. Each one was within 10ms for me when I didn't.

I'm seeing the same thing. All under 10ms and varying wildly with such small ops. I'll need to add _a lot_ more elements, too :)

One more thing...even if your method is faster in this particular case, I wonder if that's a consequence of the elements being so small? My naive guess is that as the elements get more complicated, the DOM reflows caused by all of the reflows would cause bigger delays? That's just a theory, though – I'm excited to test it further.

**Comment by jamesplease** [comment]
Test2 with 10k elements (and 10001 DOM ops)

4138.13000000664
6650.095000019064
7197.916000004625 

Test3 with 10k elements (and 1 DOM op):

2050.0130000000354
1767.1280000067782 
1961.115999991307 

# 

I must have messed up copying over the code, or something else is amiss, 'cuz it seems like the one-DOM-op method is far faster here.

# 

Update: just pushed the 10k tests to master branch on that repo.

# 

Test4: Same as test2 but with more complex list items and styling

6013.8650000153575 
9130.596999981208
10607.688999996753

Test5: Same as test3 but with more complex list items and styling

8782.682999997633 
8112.836999993306
9032.57099998882
